### PR TITLE
Include parent-imported components in implicit merge targets

### PR DESCRIPTION
**Pull request by amisevsk**
### What does this PR do?
Include components that were imported by a parent in potential targets for implicit merging of container contributions

### What issues does this PR fix or reference?
Closes https://github.com/devfile/devworkspace-operator/issues/1158

### Is it tested? How?
I haven't had time to test this PR thoroughly; it should be ensured that including parents in merge targets still behaves as expected:
* What happens if a devfile contains both a component and a parent? Is the merge processed on the "expected" component?
  * E.g. if we have a parent devfile and add e.g. a 'database' container, is merging still processed as expected?

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

**Comment by AObuchow** [comment]
> What happens if a devfile contains both a component and a parent? Is the merge processed on the "expected" component?

I'm in the process of writing test cases right now, but it seems that the parent component will be selected as the implicit merge target, i.e.

```YAML
  spec:
    parent:
      uri: parent.yaml # Parent's container component will get selected as implcit merge target
    components:
      - name: first-child-component # Won't get selected as an implicit merge target
        container:
          image: quay.io/devfile/universal-developer-image:latest
          memoryLimit: 2Gi
          mountSources: true
  
  contributions:
    - name: che-code
      uri: che-code.yaml
```

I imagine we want to prioritize implicitly merging into components defined in the "child" devworkspace, rather than the parent right?  I can try and fix this.

File: pkg/library/flatten/merge.go [hunk 1]
```diff
@@ -165,7 +165,8 @@ func needsContainerContributionMerge(flattenedSpec *dw.DevWorkspaceTemplateSpec)
 				return false, fmt.Errorf("failed to parse %s attribute on component %s as true or false", constants.MergeContributionAttribute, component.Name)
 			}
 		} else {
-			if !component.Attributes.Exists(constants.PluginSourceAttribute) {
+			pluginSource := component.Attributes.GetString(constants.PluginSourceAttribute, nil)
+			if pluginSource == "" || pluginSource == "parent" {
```
Comment (AObuchow) [inline pkg/library/flatten/merge.go:169]: It doesn't really sit well to me that the merge target found by `needsContainerContributionMerge()` and  `findMergeTarget` may be different, but I'm not sure if that affects things. Regardless, I'm going to fix this disparity shortly

**Comment by amisevsk** [comment]
> I imagine we want to prioritize implicitly merging into components defined in the "child" devworkspace, rather than the parent right? I can try and fix this.

I think the only path forward that remains consistent + configurable would be to merge into parents first; thinking through three cases:

Let's call the approach where we always try to merge into the main devfile first "Plan A" and the approach where we merge into parent first "Plan B"

* Parent devfile + no components in main devfile: No confusion here, either plan is equivalent
* Parent devfile + a component you want to merge into inside the main devfile:
  * Plan A: this is the normal behavior
  * Plan B: you would need to add the explicit `merge-contributions` attribute to your devfile's component (easy to do)
* Parent devfile + a component in main devfile; you want to merge into the parent devfile's component
  * Plan A: the only way you can do it is by modifying the parent devfile to have the `merge-contributions` attribute (hard to do; fragile as parent cannot be used in devfiles that use the `merge-contributions` attr)
  * Plan B: normal behavior; parent component will be used as merge target

**Comment by AObuchow** [comment]
@amisevsk I agree with your reasoning, good call. Will make the appropriate changes

**Comment by AObuchow** [comment]
I verified that this patch fixes the issue in https://github.com/eclipse/che/issues/22371 by installing Che with patched DWO (from this PR) and creating a workspace from this [devfile](https://github.com/AObuchow/dw-parent-contribution/blob/main/devfile.yaml). I saw that the `che-code-runtime-description` container was now being contributed to the tools container, and not being run as a standalone container:

image

**Review by AObuchow — APPROVED** [review]

**Pull request merged**